spec/p2p/v0.34/addressbook.md

```diff
@@ -0,0 +1,368 @@
+# Address Book
+
+The address book tracks information about other nodes in the network.
+
+The primary information about a node is its peer address,
+composed by a network address and an ID.
+Addresses can be manually configured (e.g., as persistent peers),
+can be learned when a peer attempts to connect to the node,
+or can be retrieved via the [Peer Exchange protocol](./pex-protocol.md).
+
+The [Peer Manager](./peer_manager.md) retrieves from the address book peer
+addresses to dial, so to establish outbound connections.
+It also records in the address book failed attempts to connect to a peer.
+
+The [Peer Exchange protocol](./pex-protocol.md) feds the address book with
+addresses received from peers, and retrieves from the address book random
+selection of addresses it provides to peers.
+
+Once the node is connected to a peer, it uses the address book to record
+information about the quality of the peer.
+The quality of a peer is currently restricted to the definition of
+[good](#good-peers) or [bad](#bad-peers) peers.
+It influences, in particular, the selection of candidate peers to dial.
+
+## New addresses
+
+The `AddAddress` method adds a new address to the address book.
+
+The added address is associated to a *source* address, which identifies the
+node from which the address was learned.
+
+Addresses are added to the address book in the following situations:
+
+1. When a peer address is received via PEX protocol, the source is the sender
+   of the PEX message
+1. When an inbound peer is added to the PEX reactor, in this case the peer
+   itself is set as the source of its address
+1. When the switch is instructed to dial addresses via the `DialPeersAsync`
+   method, in this case the node itself is set as the source
+
+An address contains a node ID and a network address (IP and port).
+The address book can keep multiple network addresses for the same node ID, but
+this number should be limited.
+A new address with a node ID that is already present in the address book is
+**not added** when:
```

Comment:
This is an example where higher level invariant should be captured, i.e., intention behind this mechanism, and not just rules, as it is not really clear why this mechanism exists and what it is trying to achieve.